[2024-03-25 11:19:14,207][flwr][WARNING] - Both server and strategy were provided, ignoring strategy
[2024-03-25 11:19:14,208][flwr][INFO] - Starting Flower simulation, config: ServerConfig(num_rounds=2, round_timeout=None)
[2024-03-25 11:19:18,997][flwr][INFO] - Flower VCE: Ray initialized with resources: {'node:__internal_head__': 1.0, 'node:127.0.0.1': 1.0, 'object_store_memory': 7776686899.0, 'memory': 15553373799.0, 'CPU': 6.0, 'GPU': 1.0}
[2024-03-25 11:19:18,998][flwr][INFO] - Optimize your simulation with Flower VCE: https://flower.dev/docs/framework/how-to-run-simulations.html
[2024-03-25 11:19:18,998][flwr][INFO] - Flower VCE: Resources for each Virtual Client: {'num_cpus': 4, 'num_gpus': 1}
[2024-03-25 11:19:19,008][flwr][INFO] - Flower VCE: Creating VirtualClientEngineActorPool with 1 actors
[2024-03-25 11:19:19,009][flwr][INFO] - Initializing global parameters
[2024-03-25 11:19:19,009][flwr][INFO] - Requesting initial parameters from one random client
[2024-03-25 11:19:24,861][flwr][INFO] - Received initial parameters from one random client
[2024-03-25 11:19:24,861][flwr][INFO] - Evaluating initial parameters
[2024-03-25 11:19:40,882][flwr][INFO] - initial parameters (loss, other metrics): 0.04536642634868622, {'accuracy': 28.61328125}
[2024-03-25 11:19:40,882][flwr][INFO] - FL starting
[2024-03-25 11:19:40,882][flwr][DEBUG] - fit_round 1: strategy sampled 5 clients (out of 10)
[2024-03-25 11:20:25,937][flwr][DEBUG] - fit_round 1 received 5 results and 0 failures
[2024-03-25 11:20:26,168][flwr][ERROR] - 'client_id'
[2024-03-25 11:20:26,168][flwr][ERROR] - Traceback (most recent call last):
  File "C:\Users\smpoulio_local\miniconda3\envs\pyt_flwr\lib\site-packages\flwr\simulation\app.py", line 308, in start_simulation
    hist = run_fl(
  File "C:\Users\smpoulio_local\miniconda3\envs\pyt_flwr\lib\site-packages\flwr\server\app.py", line 225, in run_fl
    hist = server.fit(num_rounds=config.num_rounds, timeout=config.round_timeout)
  File "C:\Users\smpoulio_local\miniconda3\envs\pyt_flwr\lib\site-packages\flwr\server\server.py", line 109, in fit
    res_fit = self.fit_round(
  File "C:\Users\smpoulio_local\miniconda3\envs\pyt_flwr\lib\site-packages\flwr\server\server.py", line 248, in fit_round
    ] = self.strategy.aggregate_fit(server_round, results, failures)
  File "C:\Users\smpoulio_local\Documents\GitHub\FlowerYtTut\strategy.py", line 110, in aggregate_fit
    temp_dict = {"client_id" : m["client_id"], "server_round": server_round, "loss": m["loss"], "accuracy": m["accuracy"], "fit_mins": m["fit_mins"]}
KeyError: 'client_id'

[2024-03-25 11:20:26,180][flwr][ERROR] - Your simulation crashed :(. This could be because of several reasons. The most common are: 
	 > Sometimes, issues in the simulation code itself can cause crashes. It's always a good idea to double-check your code for any potential bugs or inconsistencies that might be contributing to the problem. For example: 
		 - You might be using a class attribute in your clients that hasn't been defined.
		 - There could be an incorrect method call to a 3rd party library (e.g., PyTorch).
		 - The return types of methods in your clients/strategies might be incorrect.
	 > Your system couldn't fit a single VirtualClient: try lowering `client_resources`.
	 > All the actors in your pool crashed. This could be because: 
		 - You clients hit an out-of-memory (OOM) error and actors couldn't recover from it. Try launching your simulation with more generous `client_resources` setting (i.e. it seems {'num_cpus': 4, 'num_gpus': 1} is not enough for your run). Use fewer concurrent actors. 
		 - You were running a multi-node simulation and all worker nodes disconnected. The head node might still be alive but cannot accommodate any actor with resources: {'num_cpus': 4, 'num_gpus': 1}.
Take a look at the Flower simulation examples for guidance <https://flower.dev/docs/framework/how-to-run-simulations.html>.
